File: PoketDex/app/src/main/java/pt/rfsfernandes/ui/activities/MainActivity.java
package pt.rfsfernandes.ui.activities;

import android.media.MediaPlayer;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import com.google.android.material.snackbar.Snackbar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.NavDestination;
import androidx.navigation.Navigation;
import pt.rfsfernandes.MyApplication;
import pt.rfsfernandes.R;
import pt.rfsfernandes.custom.Constants;
import pt.rfsfernandes.custom.dialogs.TypeCustomDialog;
import pt.rfsfernandes.databinding.ActivityMainBinding;
import pt.rfsfernandes.model.type.Type;
import pt.rfsfernandes.viewmodels.MainViewModel;

public class MainActivity extends FragmentActivity implements NavController.OnDestinationChangedListener {
  private MainViewModel mMainViewModel;
  private ActivityMainBinding mActivityMainBinding;
  private NavController mNavControllerList;
  private NavController mNavControllerDetails;
  private MediaPlayer mMediaPlayerMenuSound;
  private MediaPlayer mMediaPlayerWalkingMusic;
  private MyApplication mMyApplication;
  private boolean showTypeInfo;
  private int pokemonId;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    mMediaPlayerMenuSound = ((MyApplication) getApplication()).getMediaPlayerMenuSound();
    mMediaPlayerWalkingMusic = ((MyApplication) getApplication()).getMediaPlayerWalkingMusic();
    mActivityMainBinding = ActivityMainBinding.inflate(getLayoutInflater());

    View view = mActivityMainBinding.getRoot();
    setContentView(view);
    mMyApplication = (MyApplication) getApplication();
    mMainViewModel = new ViewModelProvider(this).get(MainViewModel.class);

    mNavControllerList = Navigation.findNavController(this, R.id.displayPokemonList);
    mMyApplication.setLandscape(findViewById(R.id.displayPokemonDetails) != null);
    if (mMyApplication.isLandscape()) {

      mNavControllerDetails =
          Navigation.findNavController(this, R.id.displayPokemonDetails);

    }

    mNavControllerList.addOnDestinationChangedListener(this);
    initViewModel(savedInstanceState);

    mMainViewModel.isLoading(true);

    mActivityMainBinding.imageButtonSound.setOnClickListener(e -> {
      handleMusic(true);
    });

    if (mActivityMainBinding.imageButtonBack != null) {
      mActivityMainBinding.imageButtonBack.setVisibility(mMyApplication.isLandscape() ?
          View.INVISIBLE : (pokemonId != 0 ? View.VISIBLE : View.INVISIBLE));

      mActivityMainBinding.imageButtonBack.setOnClickListener(e -> {
        mNavControllerList.popBackStack();
      });
    }

  }

  private void handleMusic(boolean fromClick) {

    if (mMyApplication.isCanPlaySounds()) {
      if (fromClick) {
        mMyApplication.setCanPlaySounds(false);
        if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic.isPlaying()) {
          mMediaPlayerWalkingMusic.pause();
        }
      } else {
        if (mMediaPlayerWalkingMusic != null && !mMediaPlayerWalkingMusic.isPlaying()) {
          mMediaPlayerWalkingMusic.start();
          mMediaPlayerWalkingMusic.setLooping(true);
        }
      }

    } else {
      if (fromClick) {
        mMyApplication.setCanPlaySounds(true);
        if (mMediaPlayerWalkingMusic != null && !mMediaPlayerWalkingMusic.isPlaying()) {
          mMediaPlayerWalkingMusic.start();
          mMediaPlayerWalkingMusic.setLooping(true);
        }
      } else {
        if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic.isPlaying()) {
          mMediaPlayerWalkingMusic.pause();
        }
      }
    }

    handleVolumeButton();

  }

  public void onItemClick(int pokemonId) {
    if (!mMyApplication.isLandscape()) {
      mNavControllerList.navigate(R.id.pokemonDetailsFragment);
    } else {
      if (mActivityMainBinding.linearLaoutDetailsContainer != null) {
        mActivityMainBinding.linearLaoutDetailsContainer.setVisibility(View.VISIBLE);
      }
    }
    mMainViewModel.isLoading(true);
    mMainViewModel.pokemonById(pokemonId);

  }

  @Override
  public void onBackPressed() {
    mMyApplication.playMenuSound();

    if (!mMyApplication.isLandscape()) {
      if (mNavControllerList.getCurrentDestination() != null && mNavControllerList.getCurrentDestination().getId() == R.id.pokemonDetailsFragment) {
        mMainViewModel.getSelectedPokemonId().postValue(0);
        mNavControllerList.navigate(R.id.pokemonResultListFragment);
      } else {
        super.onBackPressed();
      }
    } else {
      super.onBackPressed();
    }

  }

  @Override
  protected void onResume() {
    super.onResume();

    handleMusic(false);
  }

  private void handleVolumeButton() {

    mActivityMainBinding.imageButtonSound.setImageDrawable(
        mMyApplication.isCanPlaySounds() ?
            ResourcesCompat.getDrawable(getResources(), R.drawable.volume_on, getTheme())
            :
            ResourcesCompat.getDrawable(getResources(), R.drawable.volume_off, getTheme())
    );

  }

  private void initViewModel(Bundle savedState) {

    mMainViewModel.getPokemonListResponseMutableLiveData().observe(this, pokemonListResponse -> {
      if (pokemonListResponse != null && pokemonListResponse.size() > 0) {
        Log.d("Info from db", "Success");
      }
    });

    mMainViewModel.getFecthErrorLiveData().observe(this, message -> {
      Snackbar.make(findViewById(android.R.id.content), message,
          Snackbar.LENGTH_LONG).show();
    });

    mMainViewModel.getSelectedPokemonId().observe(this, pokemonId -> {
      this.pokemonId = pokemonId;
      if (savedState != null && !savedState.isEmpty()) {
        if (pokemonId != 0) {

          if (!mMyApplication.isLandscape()) {
            mNavControllerList.navigate(R.id.pokemonDetailsFragment);
          } else {
            if (mActivityMainBinding.linearLaoutDetailsContainer != null) {
              mActivityMainBinding.linearLaoutDetailsContainer.setVisibility(View.VISIBLE);
            }
            mNavControllerList.navigate(R.id.pokemonResultListFragment);
          }

        } else {
          mNavControllerList.navigate(R.id.pokemonResultListFragment);
        }
        savedState.clear();
      }
    });

    mMainViewModel.getTypeMutableLiveData().observe(this, type -> {
      mMainViewModel.getSHOW_typeMutableLiveData().observe(this, show_type -> {

        if (showTypeInfo) {
          showTypeInfoDialog(type, show_type);
        }
        showTypeInfo = false;
        mMainViewModel.isLoading(false);
      });
    });

    if (savedState == null) {
      mMainViewModel.loadResults();
    }

  }

  private void showTypeInfoDialog(Type type, Constants.SHOW_TYPE show_type){
    TypeCustomDialog typeCustomDialog = new TypeCustomDialog(this);
    typeCustomDialog.show();

    switch (show_type) {
      case MOVE:
        typeCustomDialog.getPokemonTypesAdapterDoubleDamage().refreshList(type.getDamageRelations().getDoubleDamageTo());
        typeCustomDialog.getPokemonTypesAdapterHalfDamage().refreshList(type.getDamageRelations().getHalfDamageTo());
        typeCustomDialog.getPokemonTypesAdapterNoDamage().refreshList(type.getDamageRelations().getNoDamageTo());
        typeCustomDialog.getTextViewComparisonType().setText("TO");
        break;
      case POKEMON:
        typeCustomDialog.getPokemonTypesAdapterDoubleDamage().refreshList(type.getDamageRelations().getDoubleDamageFrom());
        typeCustomDialog.getPokemonTypesAdapterHalfDamage().refreshList(type.getDamageRelations().getHalfDamageFrom());
        typeCustomDialog.getPokemonTypesAdapterNoDamage().refreshList(type.getDamageRelations().getNoDamageFrom());
        typeCustomDialog.getTextViewComparisonType().setText("FROM");

        break;
    }

    typeCustomDialog.setType(type.getName());
    mMainViewModel.isLoading(false);
  }


  public void setShowTypeInfo(boolean showTypeInfo) {
    this.showTypeInfo = showTypeInfo;
  }

  @Override
  protected void onPause() {
    super.onPause();
    if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic.isPlaying()) {
      mMediaPlayerWalkingMusic.pause();
    }
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    mActivityMainBinding = null;
    if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic.isPlaying()) {
      mMediaPlayerWalkingMusic.stop();
    }

    if (mMediaPlayerWalkingMusic != null && mMediaPlayerMenuSound.isPlaying()) {
      mMediaPlayerMenuSound.stop();
    }

  }

  @Override
  public void onDestinationChanged(@NonNull NavController controller, @NonNull NavDestination destination, @Nullable Bundle arguments) {


    if (destination.getId() == R.id.pokemonDetailsFragment && !mMyApplication.isLandscape()) {
      if (mActivityMainBinding.imageButtonBack != null) {
        mActivityMainBinding.imageButtonBack.setVisibility(View.VISIBLE);
      }
    } else if (destination.getId() == R.id.pokemonResultListFragment && !mMyApplication.isLandscape()) {
      if (mActivityMainBinding.imageButtonBack != null) {
        mActivityMainBinding.imageButtonBack.setVisibility(View.GONE);
      }
    } else {
      if (mActivityMainBinding.imageButtonBack != null) {
        mActivityMainBinding.imageButtonBack.setVisibility(View.GONE);
      }
    }

  }

}

File: PoketDex/app/src/main/java/pt/rfsfernandes/pocketdex/ui/activities/MainActivity.kt
package pt.rfsfernandes.pocketdex.ui.activities

import android.media.MediaPlayer
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.FragmentActivity
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.Navigation.findNavController
import com.google.android.material.snackbar.Snackbar
import org.koin.androidx.viewmodel.ext.android.viewModel
import pt.rfsfernandes.pocketdex.MyCustomApplication
import pt.rfsfernandes.pocketdex.R
import pt.rfsfernandes.pocketdex.custom.Constants.SHOW_TYPE
import pt.rfsfernandes.pocketdex.custom.dialogs.TypeCustomDialog
import pt.rfsfernandes.pocketdex.databinding.ActivityMainBinding
import pt.rfsfernandes.pocketdex.model.service_responses.PokemonResult
import pt.rfsfernandes.pocketdex.model.type.Type
import pt.rfsfernandes.pocketdex.viewmodels.MainViewModel

class MainActivity : FragmentActivity(), NavController.OnDestinationChangedListener {
    private val mMainViewModel: MainViewModel by viewModel()
    private var mActivityMainBinding: ActivityMainBinding? = null
    private var mNavControllerList: NavController? = null
    private var mNavControllerDetails: NavController? = null
    private var mMediaPlayerMenuSound: MediaPlayer? = null
    private var mMediaPlayerWalkingMusic: MediaPlayer? = null
    private var mMyCustomApplication: MyCustomApplication? = null
    private var showTypeInfo = false
    private var pokemonId = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mMediaPlayerMenuSound = (application as MyCustomApplication).mediaPlayerMenuSound
        mMediaPlayerWalkingMusic = (application as MyCustomApplication).mediaPlayerWalkingMusic
        mActivityMainBinding = ActivityMainBinding.inflate(layoutInflater)
        val view: View = mActivityMainBinding!!.root
        setContentView(view)
        mMyCustomApplication = application as MyCustomApplication
        mNavControllerList = findNavController(this, R.id.displayPokemonList)
        mMyCustomApplication!!.isLandscape = findViewById<View?>(R.id.displayPokemonDetails) != null
        if (mMyCustomApplication!!.isLandscape) {
            mNavControllerDetails = findNavController(this, R.id.displayPokemonDetails)
        }
        mNavControllerList!!.addOnDestinationChangedListener(this)
        initViewModel(savedInstanceState)
        mMainViewModel.isLoading(true)
        mActivityMainBinding!!.imageButtonSound.setOnClickListener { e: View? -> handleMusic(true) }
        if (mActivityMainBinding!!.imageButtonBack != null) {
            mActivityMainBinding!!.imageButtonBack!!.visibility = if (mMyCustomApplication!!.isLandscape) View.INVISIBLE else if (pokemonId != 0) View.VISIBLE else View.INVISIBLE
            mActivityMainBinding!!.imageButtonBack!!.setOnClickListener { e: View? -> mNavControllerList!!.popBackStack() }
        }
    }

    private fun handleMusic(fromClick: Boolean) {
        if (mMyCustomApplication!!.canPlaySounds) {
            if (fromClick) {
                mMyCustomApplication!!.canPlaySounds = false
                if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic!!.isPlaying) {
                    mMediaPlayerWalkingMusic!!.pause()
                }
            } else {
                if (mMediaPlayerWalkingMusic != null && !mMediaPlayerWalkingMusic!!.isPlaying) {
                    mMediaPlayerWalkingMusic!!.start()
                    mMediaPlayerWalkingMusic!!.isLooping = true
                }
            }
        } else {
            if (fromClick) {
                mMyCustomApplication!!.canPlaySounds = true
                if (mMediaPlayerWalkingMusic != null && !mMediaPlayerWalkingMusic!!.isPlaying) {
                    mMediaPlayerWalkingMusic!!.start()
                    mMediaPlayerWalkingMusic!!.isLooping = true
                }
            } else {
                if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic!!.isPlaying) {
                    mMediaPlayerWalkingMusic!!.pause()
                }
            }
        }
        handleMuteButton()
    }

    fun onItemClick(pokemonId: Int) {
        if (!mMyCustomApplication!!.isLandscape) {
            mNavControllerList!!.navigate(R.id.pokemonDetailsFragment)
        } else {
            if (mActivityMainBinding!!.linearLaoutDetailsContainer != null) {
                mActivityMainBinding!!.linearLaoutDetailsContainer!!.visibility = View.VISIBLE
            }
        }
        mMainViewModel.isLoading(true)
        mMainViewModel.pokemonById(pokemonId)
    }

    override fun onBackPressed() {
        mMyCustomApplication!!.playMenuSound()
        if (!mMyCustomApplication!!.isLandscape) {
            if (mNavControllerList!!.currentDestination != null && mNavControllerList!!.currentDestination!!.id == R.id.pokemonDetailsFragment) {
                mMainViewModel.selectedPokemonId.postValue(0)
                mNavControllerList!!.navigate(R.id.pokemonResultListFragment)
            } else {
                super.onBackPressed()
            }
        } else {
            super.onBackPressed()
        }
    }

    override fun onResume() {
        super.onResume()
        handleMusic(false)
    }

    /**
     * Handles behaviour of the mute button
     */
    private fun handleMuteButton() {
        mActivityMainBinding!!.imageButtonSound.setImageDrawable(
                if (mMyCustomApplication!!.canPlaySounds) ResourcesCompat.getDrawable(resources, R.drawable.volume_on, theme) else ResourcesCompat.getDrawable(resources, R.drawable.volume_off, theme)
        )
    }

    /**
     * Initiates viewModel observers
     */
    private fun initViewModel(savedState: Bundle?) {
        mMainViewModel.pokemonListResponseMutableLiveData.observe(this) { pokemonListResponse: List<PokemonResult?>? ->
            if (pokemonListResponse != null && pokemonListResponse.size > 0) {
                Log.d("Info from db", "Success")
            }
        }
        mMainViewModel.fetchErrorLiveData.observe(this) { message: String? ->
            Snackbar.make(findViewById(android.R.id.content), message!!,
                    Snackbar.LENGTH_LONG).show()
        }
        mMainViewModel.selectedPokemonId.observe(this) { pokemonId: Int ->
            this.pokemonId = pokemonId
            if (savedState != null && !savedState.isEmpty) {
                if (pokemonId != 0) {
                    if (!mMyCustomApplication!!.isLandscape) {
                        mNavControllerList!!.navigate(R.id.pokemonDetailsFragment)
                    } else {
                        if (mActivityMainBinding!!.linearLaoutDetailsContainer != null) {
                            mActivityMainBinding!!.linearLaoutDetailsContainer!!.visibility = View.VISIBLE
                        }
                        mNavControllerList!!.navigate(R.id.pokemonResultListFragment)
                    }
                } else {
                    mNavControllerList!!.navigate(R.id.pokemonResultListFragment)
                }
                savedState.clear()
            }
        }
        mMainViewModel.typeMutableLiveData.observe(this) { type: Type ->
            mMainViewModel.sHOW_typeMutableLiveData.observe(this) { show_type: SHOW_TYPE ->
                if (showTypeInfo) {
                    showTypeInfoDialog(type, show_type)
                }
                showTypeInfo = false
                mMainViewModel.isLoading(false)
            }
        }
        if (savedState == null) {
            mMainViewModel.getPokemonList()
        }
    }

    /**
     * Shows a dialog containing the picked move/pokemon type and it's counters/weaknesses
     *
     * @param type      Type of the move/pokemon
     * @param show_type If it's a move or a pokemon
     */
    private fun showTypeInfoDialog(type: Type, show_type: SHOW_TYPE) {
        val typeCustomDialog = TypeCustomDialog(this)
        typeCustomDialog.show()
        when (show_type) {
            SHOW_TYPE.MOVE -> {
                typeCustomDialog.pokemonTypesAdapterDoubleDamage?.refreshList(type.damageRelations.doubleDamageTo)
                typeCustomDialog.pokemonTypesAdapterHalfDamage?.refreshList(type.damageRelations.halfDamageTo)
                typeCustomDialog.pokemonTypesAdapterNoDamage?.refreshList(type.damageRelations.noDamageTo)
                typeCustomDialog.textViewComparisonType?.text = "TO"
            }
            SHOW_TYPE.POKEMON -> {
                typeCustomDialog.pokemonTypesAdapterDoubleDamage?.refreshList(type.damageRelations.doubleDamageFrom)
                typeCustomDialog.pokemonTypesAdapterHalfDamage?.refreshList(type.damageRelations.halfDamageFrom)
                typeCustomDialog.pokemonTypesAdapterNoDamage?.refreshList(type.damageRelations.noDamageFrom)
                typeCustomDialog.textViewComparisonType?.text = "FROM"
            }
        }
        typeCustomDialog.setType(type.name)
        mMainViewModel.isLoading(false)
    }

    fun setShowTypeInfo(showTypeInfo: Boolean) {
        this.showTypeInfo = showTypeInfo
    }

    override fun onPause() {
        super.onPause()
        if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic!!.isPlaying) {
            mMediaPlayerWalkingMusic!!.pause()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mActivityMainBinding = null
        if (mMediaPlayerWalkingMusic != null && mMediaPlayerWalkingMusic!!.isPlaying) {
            mMediaPlayerWalkingMusic!!.stop()
        }
        if (mMediaPlayerWalkingMusic != null && mMediaPlayerMenuSound!!.isPlaying) {
            mMediaPlayerMenuSound!!.stop()
        }
    }

    override fun onDestinationChanged(controller: NavController, destination: NavDestination, arguments: Bundle?) {
        if (destination.id == R.id.pokemonDetailsFragment && !mMyCustomApplication!!.isLandscape) {
            if (mActivityMainBinding!!.imageButtonBack != null) {
                mActivityMainBinding!!.imageButtonBack!!.visibility = View.VISIBLE
            }
        } else if (destination.id == R.id.pokemonResultListFragment && !mMyCustomApplication!!.isLandscape) {
            if (mActivityMainBinding!!.imageButtonBack != null) {
                mActivityMainBinding!!.imageButtonBack!!.visibility = View.GONE
            }
        } else {
            if (mActivityMainBinding!!.imageButtonBack != null) {
                mActivityMainBinding!!.imageButtonBack!!.visibility = View.GONE
            }
        }
    }
}